tests/ui/lifetimes/mismatched-lifetime-syntaxes.rs

```diff
@@ -0,0 +1,290 @@
+            //~^ ERROR lifetime flowing from input to output with different syntax
+            ContainsLifetime(&self.0)
+        }
+    }
+}
+
+/// `impl Trait` uses lifetimes in some additional ways.
+mod impl_trait {
+    #[derive(Copy, Clone)]
+    struct ContainsLifetime<'a>(&'a u8);
+
+    fn named_ref_to_impl_trait_bound<'a>(v: &'a u8) -> impl FnOnce() + '_ {
+        //~^ ERROR lifetime flowing from input to output with different syntax
+        move || _ = v
+    }
+
+    fn named_ref_to_impl_trait_precise_capture<'a>(v: &'a u8) -> impl FnOnce() + use<'_> {
+        //~^ ERROR lifetime flowing from input to output with different syntax
+        move || _ = v
+    }
+
+    fn named_path_to_impl_trait_bound<'a>(v: ContainsLifetime<'a>) -> impl FnOnce() + '_ {
+        //~^ ERROR lifetime flowing from input to output with different syntax
+        move || _ = v
+    }
+
+    fn named_path_to_impl_trait_precise_capture<'a>(
+        v: ContainsLifetime<'a>,
+        //~^ ERROR lifetime flowing from input to output with different syntax
+    ) -> impl FnOnce() + use<'_> {
+        move || _ = v
+    }
+}
+
+/// These tests serve to exercise edge cases of the lint formatting
+mod diagnostic_output {
+    fn multiple_outputs<'a>(v: &'a u8) -> (&u8, &u8) {
+        //~^ ERROR lifetime flowing from input to output with different syntax
+        (v, v)
+    }
+}
+
+/// These usages are expected to **not** trigger the lint
+mod acceptable_uses {
+    #[derive(Copy, Clone)]
+    struct ContainsLifetime<'a>(&'a u8);
+
+    struct S(u8);
+
+    fn hidden_ref_to_hidden_ref(v: &u8) -> &u8 {
+        v
+    }
+
+    fn anonymous_ref_to_anonymous_ref(v: &'_ u8) -> &'_ u8 {
+        v
+    }
+
+    fn named_ref_to_named_ref<'a>(v: &'a u8) -> &'a u8 {
+        v
+    }
+
+    fn hidden_path_to_hidden_path(v: ContainsLifetime) -> ContainsLifetime {
+        v
+    }
+
+    fn anonymous_path_to_anonymous_path(v: ContainsLifetime<'_>) -> ContainsLifetime<'_> {
+        v
+    }
+
+    fn named_path_to_named_path<'a>(v: ContainsLifetime<'a>) -> ContainsLifetime<'a> {
+        v
+    }
+
+    fn hidden_ref_to_hidden_path(v: &u8) -> ContainsLifetime {
+        ContainsLifetime(v)
+    }
+
+    fn anonymous_ref_to_anonymous_path(v: &'_ u8) -> ContainsLifetime<'_> {
+        ContainsLifetime(v)
+    }
+
+    fn named_ref_to_named_path<'a>(v: &'a u8) -> ContainsLifetime<'a> {
+        ContainsLifetime(v)
+    }
+
+    fn hidden_path_to_hidden_ref(v: ContainsLifetime) -> &u8 {
+        v.0
+    }
+
+    fn anonymous_path_to_anonymous_ref(v: ContainsLifetime<'_>) -> &'_ u8 {
+        v.0
+    }
+
+    fn named_path_to_named_ref<'a>(v: ContainsLifetime<'a>) -> &'a u8 {
+        v.0
+    }
+
+    // These may be surprising, but ampersands count as enough of a
+    // visual indicator that a reference exists that we treat
+    // references with hidden lifetimes the same as if they were
+    // anonymous.
```

Comment:
This comment should be updated too to reflect the nomenclature that we [adopted](https://github.com/rust-lang/rust/pull/120808#issuecomment-2741166247).

Reply:
This comment is referring to the [literal syntax that the user typed](https://github.com/rust-lang/rust/blob/5fe04cbebd593c6530a21cd7bd0e8e9503b1ffe4/compiler/rustc_hir/src/hir.rs#L73-L82), so it is correct to distinguish between them here. What I suggest is extending the comment to indicate that they are treated the same and include the "elided" terminology:

> These may be surprising, but ampersands count as enough of a visual indicator that a reference exists that we treat references with hidden lifetimes the same as if they were using the anonymous lifetime. We call these two cases, along with paths using the anonymous lifetime, the "elided" style.

Reply:
We can centralize discussion of this point in https://github.com/rust-lang/rust/pull/138677#discussion_r2072353367.